Summary post validation

LeoFernanndes opened the pull request:
feat: human-friendly id
altered:
  - removal of unnecessary comments
  - removal of ObjectId validation on the query endpoints
  - addition of filename as parameter on the tasks
  - addition of filename as id for data summaries
  - change of 'not found' error messages from status_code 200 to 404

LeoFernanndes committed:
feat: data summary validation

altered:
- added pydantic validation classes for id and body in post and put routes

LeoFernanndes committed:
feat: human-frindly id

altered:
  - removal of unnecessary comments
  - removal of ObjectId validation on the query endpoints
  - addition of filename as parameter on the tasks
  - addition of filename as id for data summaries
  - change of 'not found' error messages from status_code 200 to 404

matheusfbosa commented on api/v1/database.py:
```diff
@@ -8,6 +8,7 @@
 database = client.dataquality
 data_quality_collection = database.get_collection("data_quality")
```
Seria interessante guardar o nome da coleção do MongoDB numa variável declarada como constante no Python, algo como `COLLECTION_NAME`.

Além disso, podemos utilizar variáveis de ambiente para acessar valores do MongoDB, como _string_ de conexão, `database` e `collection`. Assim, essas variáveis importantes não ficam dispersas pelo código da aplicação, mas sim em um ou mais arquivos de configuração.

Sugestão de biblioteca para fácil configuração de variáveis de ambiente e _secrets_: [Dynaconf](https://www.dynaconf.com/)

matheusfbosa commented on api/v1/database.py:
```diff
@@ -18,23 +19,38 @@ def summary_helper(summary) -> dict:
-    return summary_helper(new_student)
+
+async def add_summary(summary_data: dict) -> dict:
+    summary_data['_id'] = summary_data['filename']
```
Acho que podemos deixar o campo `_id` de um documento com sua estrutura padrão de `ObjectID` do MongoDB, sendo assim um "ID interno".

Podemos criar um outro campo identificador único e lógico na nossa aplicação (algo como _key_) para identificar uma única entidade na nossa aplicação, independente de qual banco de dados estamos utilizando.

Por exemplo, utilizando o campo _key_ para uma entidade de metadados chamada `Table` (equivalente a um dataset do Kaggle):
```json
{
  "_id": "ObjectId("6138fab705e7813e13523d32")",
  "key": "csv/olist_order_items_dataset", // "<datasource>/<table_name>"
  "type": "table",
  // [...]
}
```

Agora para uma entidade de metadados chamada `Column`:
```json
{
  "_id": "ObjectId("6138fab705e7813e13523d32")",
  "key": "csv/olist_order_items_dataset/order_id", // "<datasource>/<table_name>/<column_name>"
  "type": "column",
  // [...]
}
```

matheusfbosa replied:
Em vez de utilizar "/" como separador das hierarquias, podemos utilizar ".", ":" ou outro símbolo.

matheusfbosa commented on api/v1/routes.py:
```diff
@@ -1,13 +1,52 @@
+class CategoricalFeatureSummary(BaseModel):
+    name: str
+    size: str
+    nan_count: int
```
`nan_count` (Not a Number) ou `na_count` (Not Available)? :smile:

Esse atributo é aplicado apenas para colunas numéricas de uma tabela qualquer?
Se não for, sugiro utilizarmos `na_count` para expressar a **quantidade de missing values**.

matheusfbosa commented on api/v1/routes.py:
```diff
@@ -1,13 +1,52 @@
+    datetime_features: List[DateTimeFeatureSummary]
+
+
+def validate_objectid(id):
```
Esta função não está mais sendo utilizada.
Se voltarmos a trabalhar com `ObjectId` no campo `_id`, ela voltará a ser.
Caso contrário, podemos removê-la.

matheusfbosa commented on api/v1/routes.py:
```diff
@@ -1,13 +1,52 @@
 @router.post('/')
-async def post_summary(payload: dict):
-    summary = jsonable_encoder(payload)
+async def post_summary(data_summary: DataSummary):
```
Gostei da tipagem!
É melhor do que um `dict` genérico, né? :laughing:

matheusfbosa commented on api/v1/routes.py:
```diff
@@ -16,17 +55,24 @@ async def get_all_summaries():
     if summary:
         return summary
-    return {"detail": "Not found."}
+    raise fastapi.HTTPException(detail="Not found.", status_code=404)
```
Bom uso de HTTP Status Code! :+1:

matheusfbosa commented on data_quality/src/task/task.py:
```diff
@@ -11,6 +11,8 @@ def run(self):
         transformer = GeneralTransformer(df)
         metadata_dict = transformer.transform()
+        metadata_dict['filename'] = self.param_dict['filepath'].split("/")[-1]
+        print(metadata_dict)
```
Remover qualquer `print` após o desenvolvimento e primeiros testes.
Podemos utilizar [logging](https://realpython.com/python-logging/) para acessar valores de variáveis, indicar respostas recebidas em tempo de execução do programa.

matheusfbosa commented on api/v1/routes.py:
```diff
@@ -16,17 +55,24 @@ async def get_all_summaries():
+    return {"detail": "Not found."}
 
-@router.put("/delete-all")
+@router.post("/delete-all")
```
Podemos adicionar um prefixo a todos os nomes dos nossos `endpoints` para seguirmos convenções de design de APIs REST.

Devemos primeiro escolher um nome bem representativo do tipo de recurso que estamos operando.
Sendo nossa API um microsserviço que trabalha sobre métricas e relatórios de Data Quality, podemos nomear nosso recurso como `metrics` ou `reports`. O que acham?

Por exemplo, podemos ter algo como:
- GET `/api/v1/reports`: lista todos os reports
- GET `/api/v1/reports/{report_id}`: consulta um único report
- DELETE `/api/v1/reports`: remove todos os reports
- DELETE `/api/v1/reports/{report_id}`: remove um único report

Assim não precisamos colocar verbos em nossos endpoints (como `/api/v1/delete-all`), porque estamos utilizando os verbos HTTP (como `DELETE`) que já expressam o que a operação faz.

O que acham? Melhor, né? :smiley: 

Sugestão de leitura: [Best practices for REST API design](https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/)

matheusfbosa commented on api/v1/routes.py:
```diff
@@ -1,13 +1,52 @@
 import fastapi
```
Sugiro adicionarmos um `README.md` em `/data-discovery-cidamo/api/` com informações que respondem a perguntas comuns como:
1. Executei `docker-compose up` e agora? Onde acesso a documentação Swagger da API? Qual a URL?
2. Uma imagem diz mais que mil palavras: podemos adicionar um print da documentação gerada automaticamente pelo FastAPI nesse novo `README.md` :smile:
3. Como está organizada a API? Quais são os diretórios e o que cada um é responsável por fazer?

Exemplo de `README.md`: [link](https://github.com/tiangolo/full-stack-fastapi-postgresql)

matheusfbosa commented on api/v1/routes.py:
```diff
@@ -16,17 +55,24 @@ async def get_all_summaries():
 async def delete_all_summaries():
     await delete_summaries()
-    return {"detail": "No content."}
+    raise fastapi.HTTPException(detail="Not found.", status_code=404)
```
HTTP Status Code de sucesso são da forma 2xx.
Por exemplo, para representarmos o sucesso de uma operação DELETE, devemos retornar o HTTP Status Code `204 No Content`.

matheusfbosa commented on api/v1/routes.py:
```diff
@@ -1,13 +1,52 @@
+async def post_summary(data_summary: DataSummary):
+    summary = jsonable_encoder(data_summary)
     new_summary = await add_summary(summary)
     return new_summary
```
Devemos utilizar o HTTP Status Code `201 Created` para responder com o sucesso da operação POST.

matheusfbosa reviewed (changes requested):
Code Review do bloco `API` do projeto de Data Discovery.

chunggnuhc committed:
alteracoes no data quality no caso geral e var continuas

chunggnuhc committed:
arrumando um erro

LeoFernanndes committed:
fix: pydantic model validation

altered:
  - correction of fields to be accepted by pydantic on data summry post in order to reflect data quality job changes

LeoFernanndes committed:
fix: merge request comments

altered:
  - environment variables from now on are stored at api/ as a .env read using python-decouple
  - _id is set back to the mongo default 2-byte ObjectId as summary internal reference and for users to interact with the application, filename still works as reference
  - ObjectId validation function was removed cesidering there is no more user inputs related to this field
  - Removed unneccesary prints
  - added /summaries rootpath for resources management
  - status code for successful delete chaged to 204 with consequent no content
  - status for successful create changed to 201
